File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/collectionsheetindividualdetails/IndividualCollectionSheetDetailsPresenter.java
package com.mifos.mifosxdroid.online.collectionsheetindividualdetails;

import com.mifos.api.DataManager;
import com.mifos.api.GenericResponse;
import com.mifos.api.datamanager.DataManagerCollectionSheet;
import com.mifos.api.model.IndividualCollectionSheetPayload;
import com.mifos.mifosxdroid.base.BasePresenter;
import com.mifos.objects.accounts.loan.PaymentTypeOptions;
import com.mifos.objects.collectionsheet.ClientCollectionSheet;
import com.mifos.objects.collectionsheet.LoanAndClientName;
import com.mifos.objects.collectionsheet.LoanCollectionSheet;
import com.mifos.utils.MFErrorParser;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import retrofit2.adapter.rxjava.HttpException;
import rx.Observable;
import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Action1;
import rx.plugins.RxJavaPlugins;
import rx.schedulers.Schedulers;
import rx.subscriptions.CompositeSubscription;

/**
 * Created by aksh on 20/6/18.
 */

public class IndividualCollectionSheetDetailsPresenter
        extends BasePresenter<IndividualCollectionSheetDetailsMvpView> {

    private DataManager mDataManager;
    private DataManagerCollectionSheet mDataManagerCollection;
    private CompositeSubscription mSubscription;


    @Inject
    IndividualCollectionSheetDetailsPresenter(DataManager manager,
                                          DataManagerCollectionSheet managerCollectionSheet) {
        mDataManager = manager;
        mDataManagerCollection = managerCollectionSheet;
        mSubscription = new CompositeSubscription();
    }

    @Override
    public void attachView(IndividualCollectionSheetDetailsMvpView mvpView) {
        super.attachView(mvpView);
    }

    @Override
    public void detachView() {
        super.detachView();
        if (mSubscription != null) {
            mSubscription.unsubscribe();
        }
    }

    void submitIndividualCollectionSheet(IndividualCollectionSheetPayload
                                                 individualCollectionSheetPayload) {
        checkViewAttached();
        getMvpView().showProgressbar(true);
        mSubscription.add(mDataManagerCollection
                .saveIndividualCollectionSheet(individualCollectionSheetPayload)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(new Subscriber<GenericResponse>() {
                    @Override
                    public void onCompleted() {

                    }

                    @Override
                    public void onError(Throwable e) {
                        getMvpView().showProgressbar(false);
                        try {
                            if (e instanceof HttpException) {
                                String errorMessage = ((HttpException) e).response().errorBody()
                                        .string();
                                getMvpView().showError(MFErrorParser.parseError(errorMessage)
                                        .getErrors().get(0).getDefaultUserMessage());
                            }
                        } catch (Throwable throwable) {
                            RxJavaPlugins.getInstance().getErrorHandler().handleError(e);
                        }
                    }

                    @Override
                    public void onNext(GenericResponse genericResponse) {
                        getMvpView().showProgressbar(false);
                        getMvpView().showSuccess();
                    }
                }));
    }

    List<String> filterPaymentTypeOptions(List<PaymentTypeOptions> paymentTypeOptionsList) {
        final List<String> paymentList = new ArrayList<>();
        Observable.from(paymentTypeOptionsList)
                .subscribe(new Action1<PaymentTypeOptions>() {
                    @Override
                    public void call(PaymentTypeOptions paymentTypeOption) {
                        paymentList.add(paymentTypeOption.getName());
                    }
                });
        return paymentList;
    }

    List<LoanAndClientName> filterLoanAndClientNames(List<ClientCollectionSheet>
                                                             clientCollectionSheets) {
        final List<LoanAndClientName> loansAndClientNames = new ArrayList<>();
        Observable.from(clientCollectionSheets)
                .subscribe(new Action1<ClientCollectionSheet>() {
                    @Override
                    public void call(ClientCollectionSheet clientCollectionSheet) {
                        if (clientCollectionSheet.getLoans() != null) {
                            for (LoanCollectionSheet loanCollectionSheet :
                                    clientCollectionSheet.getLoans()) {
                                loansAndClientNames.add(new
                                        LoanAndClientName(loanCollectionSheet,
                                        clientCollectionSheet.getClientName(),
                                        clientCollectionSheet.getClientId()));
                            }
                        }
                    }
                });
        return loansAndClientNames;
    }
}

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/collectionsheetindividualdetails/IndividualCollectionSheetDetailsPresenter.kt
package com.mifos.mifosxdroid.online.collectionsheetindividualdetails

import com.mifos.api.DataManager
import com.mifos.api.GenericResponse
import com.mifos.api.datamanager.DataManagerCollectionSheet
import com.mifos.api.model.IndividualCollectionSheetPayload
import com.mifos.mifosxdroid.base.BasePresenter
import com.mifos.objects.accounts.loan.PaymentTypeOptions
import com.mifos.objects.collectionsheet.ClientCollectionSheet
import com.mifos.objects.collectionsheet.LoanAndClientName
import com.mifos.utils.MFErrorParser
import retrofit2.HttpException
import rx.Observable
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.plugins.RxJavaPlugins
import rx.schedulers.Schedulers
import rx.subscriptions.CompositeSubscription
import java.util.*
import javax.inject.Inject

/**
 * Created by aksh on 20/6/18.
 */
class IndividualCollectionSheetDetailsPresenter @Inject internal constructor(private val mDataManager: DataManager,
                                                                             private val mDataManagerCollection: DataManagerCollectionSheet) : BasePresenter<IndividualCollectionSheetDetailsMvpView?>() {
    private val mSubscription: CompositeSubscription?
    override fun attachView(mvpView: IndividualCollectionSheetDetailsMvpView?) {
        super.attachView(mvpView)
    }

    override fun detachView() {
        super.detachView()
        mSubscription?.unsubscribe()
    }

    fun submitIndividualCollectionSheet(individualCollectionSheetPayload: IndividualCollectionSheetPayload?) {
        checkViewAttached()
        mvpView!!.showProgressbar(true)
        mSubscription!!.add(mDataManagerCollection
                .saveIndividualCollectionSheet(individualCollectionSheetPayload)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(object : Subscriber<GenericResponse?>() {
                    override fun onCompleted() {}
                    override fun onError(e: Throwable) {
                        mvpView!!.showProgressbar(false)
                        try {
                            if (e is HttpException) {
                                val errorMessage = e.response()?.errorBody()
                                        ?.string()
                                mvpView!!.showError(MFErrorParser.parseError(errorMessage)
                                        .errors[0].defaultUserMessage)
                            }
                        } catch (throwable: Throwable) {
                            RxJavaPlugins.getInstance().errorHandler.handleError(e)
                        }
                    }

                    override fun onNext(genericResponse: GenericResponse?) {
                        mvpView!!.showProgressbar(false)
                        mvpView!!.showSuccess()
                    }
                }))
    }

    fun filterPaymentTypeOptions(paymentTypeOptionsList: List<PaymentTypeOptions>?): List<String> {
        val paymentList: MutableList<String> = ArrayList()
        Observable.from(paymentTypeOptionsList)
                .subscribe { paymentTypeOption -> paymentTypeOption.name?.let { paymentList.add(it) } }
        return paymentList
    }

    fun filterLoanAndClientNames(clientCollectionSheets: List<ClientCollectionSheet>?): List<LoanAndClientName> {
        val loansAndClientNames: MutableList<LoanAndClientName> = ArrayList()
        Observable.from(clientCollectionSheets)
                .subscribe { clientCollectionSheet ->
                    if (clientCollectionSheet.loans != null) {
                        for (loanCollectionSheet in clientCollectionSheet.loans!!) {
                            loansAndClientNames.add(LoanAndClientName(loanCollectionSheet,
                                    clientCollectionSheet.clientName,
                                    clientCollectionSheet.clientId))
                        }
                    }
                }
        return loansAndClientNames
    }

    init {
        mSubscription = CompositeSubscription()
    }
}